Decode typed String dictionaries

yonaskolb opened the pull request:
This adds support for decoding typed dictionaries with a String key, specifically:
`[String: JSONObjectConvertible]`
`[String: JSONRawType]`
`[String: JSONPrimitiveConvertible]`

Includes optional variants and tests

Commit:
Decode typed String dictionaries

lucianomarisi commented:
Thanks for the PR. 

Would it be possible to add some unit tests for the `[String: JSONPrimitiveConvertible]` and `[String: JSONPrimitiveConvertible]?` cases? Also could you rebase this branch onto the latest master since the branch is out-of-date?

Out of curiosity really, I never had a need to use typed dictionaries, what's the use case for this?

yonaskolb commented:
Rebased off master, added JSONPrimitiveConvertible tests, and squashed commits 👍

Typed dictionaries are for json structures where the keys are not known up front. For example in this json http://petstore.swagger.io/v2/swagger.json there are a variable number of keys under `paths` such as `paths./pet`, so we can create a [String: Endpoint] dictionary where the key is the path
